src/vaev-engine/style/rules.cpp

```diff
@@ -257,4 +257,296 @@ void MediaRule::repr(Io::Emit& e) const {
+    }
+
+    MatchingRules _mergeMatchedRules(Vec<Cursor<Tuple<usize, Cursor<StyleRule>>>>&& cursors, Gc::Ref<Dom::Element> el) {
+        MatchingRules matchingRules;
+
+        usize countMatchesWithCurrentRule = 0;
+        usize lastRuleId = 0;
+        Cursor<StyleRule> lastStyleRule = nullptr;
+
+        auto maybeFinalizeNfixOrRule = [&]() {
+            if (not lastStyleRule)
+                return;
+
+            if (auto nfix = lastStyleRule->selector.is<Nfix>()) {
+                if (nfix->type != Nfix::OR)
+                    return;
+
+                if (countMatchesWithCurrentRule == 1) {
+                    _evalStyleRule(*lastStyleRule, el, matchingRules);
+                } else {
+                    matchingRules.pushBack({lastStyleRule, spec(lastStyleRule->selector)});
+                }
+            }
+        };
+
+        while (cursors.len() > 0) {
+            usize bestCursorIdx = 0;
+            for (usize i = 1; i < cursors.len(); i++) {
+                if (cursors[i]->v0 < cursors[bestCursorIdx]->v0) {
+                    bestCursorIdx = i;
+                }
+            }
+
+            auto& styleRule = *cursors[bestCursorIdx]->v1;
+
+            // NOTE: This is quite hot code and doing this check every time is not ideal, but it was the only way found to
+            // allow defering the evaluation of OR infixes until we know how many times this rule was matched.
+            if (lastRuleId != cursors[bestCursorIdx]->v0) {
+                maybeFinalizeNfixOrRule();
+                countMatchesWithCurrentRule = 1;
+            } else {
+                countMatchesWithCurrentRule++;
+            }
+
+            if (RuleLookup::lookupIsMatch(styleRule.selector)) {
+                matchingRules.pushBack({&styleRule, spec(styleRule.selector)});
+            } else if (auto nfix = styleRule.selector.is<Nfix>()) {
+                if (nfix->type == Nfix::AND) {
+                    auto neededCount = ruleIdToNeededCount.tryGet(cursors[bestCursorIdx]->v0);
+
+                    if (not neededCount) {
+                        _evalStyleRule(styleRule, el, matchingRules);
+                    } else if (countMatchesWithCurrentRule == *neededCount) {
+                        if (nfix->inners.len() == countMatchesWithCurrentRule) {
+                            matchingRules.pushBack({&styleRule, spec(styleRule.selector)});
+                        } else {
+                            _evalStyleRule(styleRule, el, matchingRules);
+                        }
+                    }
+                } else if (nfix->type == Nfix::OR) {
+                    // Do nothing.
+                    // Deffering the evaluation to after we know how many times this rule was matched.
+                } else {
+                    _evalStyleRule(styleRule, el, matchingRules);
+                }
+            } else {
+                _evalStyleRule(styleRule, el, matchingRules);
+            }
```

Comment:
Seeing `_evalStyleRule(styleRule, el, matchingRules);` in multiple leafs of the code tells me this code can be refactored